Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -0,0 +1,825 @@
+        if (!pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesURL(url);
+    };
+
+    for (auto& deniedPermissionEntry : deniedPermissionMatchPatterns) {
+        if (urlMatchesWildcardHostPatterns(deniedPermissionEntry.key))
+            return cacheResultAndReturn(PermissionState::DeniedImplicitly);
+    }
+
+    for (auto& grantedPermissionEntry : grantedPermissionMatchPatterns) {
+        if (urlMatchesWildcardHostPatterns(grantedPermissionEntry.key))
+            return cacheResultAndReturn(PermissionState::GrantedImplicitly);
+    }
+
+    // Finally, check for requested patterns, allowing any pattern that matches. This is the default state
+    // of the extension before any patterns are granted or denied, so it should always be last.
+
+    if (skipRequestedPermissions)
+        return cacheResultAndReturn(PermissionState::Unknown);
+
+    auto requestedMatchPatterns = m_extension->allRequestedMatchPatterns();
+    for (auto& requestedMatchPattern : requestedMatchPatterns) {
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(requestedMatchPattern))
+            return cacheResultAndReturn(PermissionState::RequestedExplicitly);
+
+        if (urlMatchesWildcardHostPatterns(requestedMatchPattern))
+            return cacheResultAndReturn(PermissionState::RequestedImplicitly);
+    }
+
+    if (options.contains(PermissionStateOptions::RequestedWithTabsPermission) && hasPermission(_WKWebExtensionPermissionTabs, tab, options))
+        return PermissionState::RequestedImplicitly;
+
+    return cacheResultAndReturn(PermissionState::Unknown);
+}
+
+WebExtensionContext::PermissionState WebExtensionContext::permissionState(WebExtensionMatchPattern& pattern, _WKWebExtensionTab *tab, OptionSet<PermissionStateOptions> options)
+{
+    if (!pattern.isValid())
+        return PermissionState::Unknown;
+
+    if (pattern.matchesURL(baseURL()))
+        return PermissionState::GrantedImplicitly;
+
+    if (!pattern.matchesAllURLs() && !WebExtensionMatchPattern::validSchemes().contains(pattern.scheme()))
+        return PermissionState::Unknown;
+
+    if (tab && [[m_temporaryTabPermissionMatchPatterns objectForKey:tab] matchesPattern:pattern.wrapper()])
+        return PermissionState::GrantedExplicitly;
+
+    // Access the maps here to remove any expired entries, and only do it once for this call.
+    auto& grantedPermissionMatchPatterns = this->grantedPermissionMatchPatterns();
+    auto& deniedPermissionMatchPatterns = this->deniedPermissionMatchPatterns();
+
+    // First, check for patterns that are specific to certain domains, ignoring wildcard host patterns that
+    // match all hosts. The order is denied, then granted. This makes sure denied takes precedence over granted.
+
+    auto urlMatchesPatternIgnoringWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesPattern(pattern);
```

Comment:
What will happen with `pattern` as the block argument vs `pattern` as the function argument here?

Reply:
I'll rename the arg.